### PR TITLE
fix(crd): fix multi-resource importing of CRDs

### PR DESCRIPTION
**Pull request by campionfellin**
Changes CRD importing to allow for CRDs with multiple resources / ApiObjects within them.

Fixes https://github.com/awslabs/cdk8s/issues/77

Signed-off-by: campionfellin <campionfellin@gmail.com>

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 1]
```diff
@@ -98,11 +73,48 @@ export class ImportCustomResourceDefinition extends ImportBase {
...
+  private readonly CRDs: CustomResourceDefinition[] = [];
+  
+  constructor(manifest: CustomResourceApiObject[]) {
+    super();
+
+    this.CRDs = manifest?.map((apiObject: CustomResourceApiObject) => new CustomResourceDefinition(apiObject));
+  }
+
+  public get moduleName() {
+    return MODULE_NAME.toLocaleLowerCase();
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/crd.ts]: I think we should generate a separate file for each CRD with module name based on the kind and not a single file for all of them (as if each CRD was in a separate manifest).

File: examples/crd/.gitignore [hunk 2]
```diff
@@ -1,5 +1,4 @@
 *.d.ts
 *.js
 node_modules
-dist/
-imports/
```
[the remark this answers is not in the document]
Comment (campionfellin) [reply examples/crd/.gitignore]: Yep, wasn't sure the resolution of this one: https://github.com/awslabs/cdk8s/issues/51

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 3]
```diff
@@ -98,11 +73,48 @@ export class ImportCustomResourceDefinition extends ImportBase {
...
+  public static async match(source: string): Promise<undefined | CustomResourceApiObject[]> {
+    let manifest;
+    if (source.startsWith('https://')) {
+      manifest = yaml.parseAllDocuments(await httpsGet(source));
+    } else if (path.extname(source) === '.yaml' || path.extname(source) === '.yml' || path.extname(source) === '.json') {
+      if (!(await fs.pathExists(source))) {
+        throw new Error(`can't find file ${source}`);
+      }
+
+      manifest = yaml.parseAllDocuments(await fs.readFile(source, 'utf-8'));
```
[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/crd.ts]: Done!

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 4]
```diff
@@ -98,11 +73,48 @@ export class ImportCustomResourceDefinition extends ImportBase {
...
+    }
+
+    if (!manifest) {
+      return undefined;
+    }
+
+    return manifest?.map((doc: yaml.ast.Document) => doc.toJSON())
```
[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/crd.ts]: Done!

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 5]
```diff
@@ -98,11 +73,48 @@ export class ImportCustomResourceDefinition extends ImportBase {
...
+  }
+
+  private readonly CRDs: CustomResourceDefinition[] = [];
+  
+  constructor(manifest: CustomResourceApiObject[]) {
+    super();
+
+    this.CRDs = manifest?.map((apiObject: CustomResourceApiObject) => new CustomResourceDefinition(apiObject));
```
[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/crd.ts]: Done!

File: packages/cdk8s-cli/lib/import/crd.ts (hunk 1 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/crd.ts]: Done!

File: packages/cdk8s-cli/lib/import/base.ts [hunk 6]
```diff
@@ -19,33 +19,38 @@ export interface ImportOptions {
...
+      code.indentation = 2;
+      await this.generateTypeScript(code, name);
+      code.closeFile(fileName);
+    
+      if (isTypescript) {
+        await code.save(outdir);
+      } 
+    }
+
+    if (isTypescript) return
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/base.ts]:
```suggestion
    if (isTypescript) return;
```

File: packages/cdk8s-cli/lib/import/base.ts [hunk 7]
```diff
@@ -19,33 +19,38 @@ export interface ImportOptions {
...
       await code.save('.');
-      await jsiiCompile('.', {
-        main: this.moduleName,
-        name: this.moduleName,
-      });
 
+      await this.moduleNames.forEach(name => jsiiCompile('.', {
+        main: name,
+        name,
+      }))
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/base.ts]: I rather avoid `forEach` in async code:

```suggestion
for (const moduleName of this.moduleNames) {
  await jsiiCompile('.', { main: moduleName, name: moduleName });
}
```

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 8]
```diff
@@ -106,6 +83,44 @@ export class ImportCustomResourceDefinition extends ImportBase {
...
+
+      manifest = await fs.readFile(source, 'utf-8');
+    }
+
+  
+    if (!manifest) {
+      return undefined;
+    }
+
+    return yaml.parseAllDocuments(manifest).map((doc: yaml.ast.Document) => doc.toJSON());
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/crd.ts]: I feel like we should filter only the CustomResourceDefinition objects from the manifest, no? Also, add a test that verifies how this behaves when the manifest has more than just a CRD.

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 9]
```diff
@@ -106,6 +83,44 @@ export class ImportCustomResourceDefinition extends ImportBase {
...
+  }
+
+  private readonly CRDs: CustomResourceDefinition[] = [];
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/crd.ts]: name should be camel case

File: packages/cdk8s-cli/lib/import/base.ts (hunk 7 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/base.ts]: Yeah learned something new about async `forEach`s while doing this. Missed one, will fix it!

File: packages/cdk8s-cli/lib/import/crd.ts (hunk 8 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/crd.ts]: I agree. Would we want some sort of message to users that we filtered out some objects from their manifest?

File: packages/cdk8s-cli/lib/import/base.ts (hunk 6 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/base.ts]: Fixed!

File: packages/cdk8s-cli/lib/import/crd.ts (hunk 9 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/crd.ts]: Fixed!

File: packages/cdk8s-cli/test/import/__snapshots__/import-crd.test.js.snap [hunk 10]
```diff
@@ -12767,3 +12767,494 @@ If unset, the Kubelet will not modify the ownership and permissions of any volum
   "version": "0.0.0",
 }
 `;
+
+<<<<<<< HEAD
```
Comment (eladb) [inline packages/cdk8s-cli/test/import/__snapshots__/import-crd.test.js.snap]: Resolve

File: packages/cdk8s-cli/tsconfig.json [hunk 11]
```diff
@@ -10,7 +10,7 @@
       "es2018"
     ],
     "module": "CommonJS",
-    "noEmitOnError": false,
+    "noEmitOnError": true,
```
Comment (eladb) [inline packages/cdk8s-cli/tsconfig.json]: We want to emit on error. Revert please

File: packages/cdk8s-cli/tsconfig.json [hunk 12]
```diff
@@ -28,8 +28,7 @@
     "**/*.ts"
   ],
   "exclude": [
-    "node_modules",
-    "templates/**/*.ts"
```
Comment (eladb) [inline packages/cdk8s-cli/tsconfig.json]: Why?

Comment (campionfellin) [reply packages/cdk8s-cli/tsconfig.json]: Ah seems JSII re-generated this when I was doing some manual testing. Will revert.

**Pull request merged**